File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/flyweight/solution/SingleJourneyFlyweightFactory.m
% SingleJourneyFlyweightFactory.m

classdef SingleJourneyFlyweightFactory < handle
	properties (GetAccess = private, SetAccess = private)
		Map % DpMap
	end

	methods
		% constructor:
		function theSingleJourneyFlyweightFactory = SingleJourneyFlyweightFactory
			theSingleJourneyFlyweightFactory.Map = DpMap;
		end

		function result = getFlyweightFor(theSingleJourneyFlyweightFactory, aAddress)
			if ~theSingleJourneyFlyweightFactory.Map.contains(aAddress)
				theSingleJourneyFlyweightFactory.Map.put(aAddress, SingleJourney(aAddress));
			end

			result = theSingleJourneyFlyweightFactory.Map.get(aAddress);
		end
	end
end

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/flyweight/lab/single_journey_flyweight_factory.cpp
// single_journey_flyweight_factory.cpp
#include <iostream>
#include "single_journey_flyweight_factory.h"
#include "single_journey.h"

single_journey_flyweight_factory::single_journey_flyweight_factory()
{}

single_journey* single_journey_flyweight_factory::get_flyweight_for(const std::string& a_address)
{
    if (a_address.empty())
    {
        dp_exception("a_address is empty!");
    }
    else 
    {
        if (my_map.contains(a_address))
        {
            auto result = my_map.get(a_address);
            return result;
        }
        auto new_journey = new single_journey(a_address);
        my_map.put(a_address, new_journey);
        return new_journey;
    }
}
